nextcloud: add recognize app

beardhatcode opened the pull request:
## Description of changes

Cleans up the way nextcloud(28|29|30)Packages etc is defined. And adding a package for recognize with support for libtensorflow (darwin suport and GPU stuff is for another PR by someone who has those kinds of machines).

On my x86 linux it now says:

> :heavy_check_mark:  The machine learning models have been downloaded successfully. 

>  :heavy_check_mark: Node.js v20.14.0 binary was installed successfully. 

 >  :heavy_check_mark: Libtensorflow was loaded successfully into Node.js. 

And it has
 >  :heavy_check_mark: Face recognition is working. 
 >  :heavy_check_mark: Object recognition is working.
 >  :heavy_check_mark: Audio recognition is working.
 >  :heavy_check_mark: Video recognition is working. 

<details><summary>old intro</summary>
Alters the way nextcloud28Packages etc is defined to allow adding custom packages.

In this case to add the recognize app which downloads the ML models at runtime (which is not compatible with the read only filesystem). Additionally, I also patch in the version of nodejs (in a very dirty way, for now). I tested it and it says

> :heavy_check_mark:  The machine learning models have been downloaded successfully. 

and 

>  :heavy_check_mark: Node.js v20.14.0 binary was installed successfully. 

but also

> :warning:  Could not load libtensorflow in Node.js. You can try to manually install libtensorflow or run in WASM mode. 

**This is mostly a request for comments on of how a thing like this can be done**. So, @Ma27, @dotlambda, @onny, and @pyrox0 (the last 4 editors of the Nextcloud packages folder), what do you think?
</details>

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

beardhatcode committed:
nextcloud: add recognize app for nextcloud

teutat3s commented:
Related PR: https://github.com/NixOS/nixpkgs/pull/211631

SuperSandro2000 commented on pkgs/servers/nextcloud/packages/default.nix:
```diff
@@ -2,10 +2,10 @@
+{ lib, pkgs, newScope, ncVersion }:
 let packages = self:
   let
+    apps = lib.importJSON (../servers/nextcloud/packages/. + (builtins.toPath "${ncVersion}.json"));
```
```suggestion
    apps = lib.importJSON (../servers/nextcloud/packages/ + "${ncVersion}.json");
```
IIRC

SuperSandro2000 commented on pkgs/servers/nextcloud/packages/adapted/nextcloud-app-recognize.nix:
```diff
@@ -0,0 +1,51 @@
+  fetchurl,
+  lib,
+  nodejs,
+  ...
```
```suggestion
```

SuperSandro2000 commented on pkgs/servers/nextcloud/packages/adapted/nextcloud-app-recognize.nix:
```diff
@@ -0,0 +1,51 @@
+    tar -xzpf "${builtins.elemAt srcs 0}" recognize;
+    tar -xzpf "${builtins.elemAt srcs 1}" recognize-${version}/models;
+    mv recognize-${version}/models recognize
+    sed -i "/'node_binary'/s:'""':'${nodejs}/bin/node':" recognize/lib/Service/SettingsService.php
```
Please use substituteInPlace

SuperSandro2000 commented on pkgs/servers/nextcloud/packages/adapted/nextcloud-app-recognize.nix:
```diff
@@ -0,0 +1,51 @@
+  meta = with lib; {
+    license = licenses.agpl3Only;
+    maintainers = with maintainers; [ beardhatcode ];
+    longDescription = ''
```
We always must add a description

SuperSandro2000 commented:
This means we don't need something like https://github.com/NuschtOS/nixos-modules/blob/da5286bc062adee0e0aaf2bd3b784b477c623422/modules/nextcloud.nix#L172-L183 but still https://github.com/NuschtOS/nixos-modules/blob/da5286bc062adee0e0aaf2bd3b784b477c623422/modules/nextcloud.nix#L210-L217 ?

Ma27 reviewed (dismissed):
> This means we don't need something like https://github.com/NuschtOS/nixos-modules/blob/da5286bc062adee0e0aaf2bd3b784b477c623422/modules/nextcloud.nix#L172-L183 but still https://github.com/NuschtOS/nixos-modules/blob/da5286bc062adee0e0aaf2bd3b784b477c623422/modules/nextcloud.nix#L210-L217 ?

To be clear: I'm very, very :-1: on doing _any_ build steps at runtime.

I expect the JS code to be built correctly and to have the settings patched by this package to point to other Nix packages (I.e. node, ffmpeg etc.). I looked through those super-questionable[0] "migration" "steps"[1] and I really don't want to rely on any of that.

[0] to say the least
[1] https://github.com/nextcloud/recognize/blob/main/lib/Migration/InstallDeps.php

beardhatcode commented on pkgs/servers/nextcloud/packages/default.nix:
```diff
@@ -23,6 +23,9 @@ let packages = self:
+
   } // lib.mapAttrs (type: pkgs:
     lib.makeExtensible (_: lib.mapAttrs (pname: data: self.mkNextcloudDerivation { inherit pname; inherit data; }) pkgs))
     generatedJson;
```
The problem is that the package in the apps.json of nextcloud expects that it is in a writable filesystem. And then at first run it downloads some more stuff, the 2 sources need to be aligned, therefore I would keep it out of the apps.json (also making it clear that it is a special case app).

SuperSandro2000 replied:
We should download those files at build time if possible.
We can verify if the right version was downloaded with grep or so.

beardhatcode commented:
> This means we don't need something like https://github.com/NuschtOS/nixos-modules/blob/da5286bc062adee0e0aaf2bd3b784b477c623422/modules/nextcloud.nix#L172-L183 but still https://github.com/NuschtOS/nixos-modules/blob/da5286bc062adee0e0aaf2bd3b784b477c623422/modules/nextcloud.nix#L210-L217 ?

I was planning to wrap the `nodejs` to add those packages so they do not need to be in the module. This way it "just works" if you enable this. But I'm still figuring out how to do this nicely.

onny commented:
I'm also going to add a third party app to the package set https://github.com/NixOS/nixpkgs/pull/328634

SuperSandro2000 commented:
> To be clear: I'm very, very 👎 on doing _any_ build steps at runtime.

I don't use declerative nextcloud apps, so I must do it like this.

So we want to patch the app before in nextcloudApps.

Ma27 commented:
> I don't use declerative nextcloud apps, so I must do it like this.

That's OK, my comment was strictly about what I'd like to see (or not see) in nixpkgs.

> So we want to patch the app before in nextcloudApps.

Most likely yes.

onny commented:
We managed to package third-party apps in https://github.com/NixOS/nixpkgs/pull/328634

So you can add your derivation into `pkgs/servers/nextcloud/packages/thirdparty.nix`

beardhatcode commented:
> We managed to package third-party apps in #328634
> 
> So you can add your derivation into `pkgs/servers/nextcloud/packages/thirdparty.nix`


Rebased on that, I'm still unable to get the tfjs stuff to compile under nix as this thing seems to insist on downloading stuff. So I'm still stuck on:

>  Could not load libtensorflow in Node.js. You can try to manually install libtensorflow or run in WASM mode. 


If anyone knows how to fix it, that would be much appreciated.

Ma27 commented:
> This app is not really third-party however.

Yeah, we should probably change the naming.
What I'm wondering is, would it be possible to track the version/src via nc4nix and only do overrides in this file?

beardhatcode commented:
Yea, I think that should be possible. I'd like to first get the tfjs node thing working. Because atm, what I have looks horrible, and it does not even work:

```nix
{
  stdenv,
  fetchurl,
  lib,
  nodejs,
  python311,
  python311Packages,

  util-linux, # needed?
  ...
}:
stdenv.mkDerivation rec {
  pname = "nextcloud-app-recognise";
  version = "6.0.1";

  srcs = [
    (fetchurl {
      inherit version;
      url = "https://github.com/nextcloud/recognize/releases/download/v${version}/recognize-${version}.tar.gz";
      hash = "sha256-HQHfcQY2N6INLfaLDWbWn0SSzvkSZqL0rPl2koqej6c=";
    })

    (fetchurl {
      inherit version;
      url = "https://github.com/nextcloud/recognize/archive/refs/tags/v${version}.tar.gz";
      hash = "sha256-4iJVKz/HfzIUouiOK/NUkJUm4+p9meVptQwJ4OG1JaU=";
    })
    (fetchurl {
      inherit version;
      url = "https://storage.googleapis.com/tensorflow/libtensorflow/libtensorflow-cpu-linux-x86_64-2.9.1.tar.gz";
      hash = "sha256-f1ENJUbj214QsdEZRjaJAD1YeEKJKtPJW8pRz4KCAXM=";
    })
    (fetchurl {
      inherit version;
      url = "https://nodejs.org/download/release/v20.15.1/node-v20.15.1-headers.tar.gz";
      hash = "sha256-jCMFxt9dFFJeBxHw2jgpVgCYffTCcQxzjAFACGKhdrQ=";
    })

  ];

  unpackPhase = ''
    tar -xzpf "${builtins.elemAt srcs 0}" recognize;
    tar -xzpf "${builtins.elemAt srcs 1}" recognize-${version}/models;
    mv recognize-${version}/models recognize
    #cd recognize/node_modules/@tensorflow/tfjs-node/deps
    #tar -xzpf "${builtins.elemAt srcs 2}"
    #cd -
    cd recognize/node_modules/@tensorflow/tfjs-node/lib
    tar -xzpf "${builtins.elemAt srcs 3}"
    cd -

    # recognize/node_modules/@tensorflow/tfjs-node/deps/lib/libtensorflow.so.2.9.1

  '';

  buildInputs = [
    nodejs
    nodejs.pkgs.node-pre-gyp
    nodejs.pkgs.node-gyp
    python311
    util-linux
  ];
  buildPhase = ''
    mkdir tmphome
    export HOME="$(realpath tmphome)"
    sed -i "/'node_binary'/s:'""':'${nodejs}/bin/node':" recognize/lib/Service/SettingsService.php
    cd recognize
    set -x
    #ln -sf -t node_modules/@tensorflow/tfjs-node/deps ${python311Packages.tensorflow}/*
    ls node_modules/@tensorflow/tfjs-node/deps
    export CPPFLAGS="-I${nodejs}/include/node -Ideps/include -I${python311Packages.tensorflow}/lib/python3.11/site-packages/tensorflow/include"
    cd node_modules/@tensorflow/tfjs-node
    echo "===="
    ls deps/include
    echo "===="
    node-pre-gyp install --prefer-offline --build-from-source --nodedir=${nodejs}/include/node
    cd -
    #npm rebuild --prefer-offline @tensorflow/tfjs-node --build-addon-from-source
    ${nodejs}/bin/node src/test_libtensorflow.js
    set +x
    exit 1
    cd -
  '';
  installPhase = ''
    set -x
    approot="$(dirname $(dirname $(find -path '*/appinfo/info.xml' | head -n 1)))"
    if [ -d "$approot" ];
    then
      mv "$approot/" $out
      chmod -R a-w $out
    fi
  '';

  meta = with lib; {
    license = licenses.agpl3Only;
    maintainers = with maintainers; [ beardhatcode ];
    longDescription = ''
      Nextcloud app that does Smart media tagging and face recognition with on-premises machine learning models.
      This app goes through your media collection and adds fitting tags, automatically categorizing your photos and music.
    '';
    homepage = "https://apps.nextcloud.com/apps/recognize";
  };
}
```

teutat3s commented:
@beardhatcode I assume you already saw the [hacky workaround in this comment](https://github.com/NixOS/nixpkgs/pull/211631#issuecomment-1398334152)? If not, I figured it might be useful to make the  node tfjs thing build.

beardhatcode commented:
> @beardhatcode I assume you already saw the [hacky workaround in this comment](https://github.com/NixOS/nixpkgs/pull/211631#issuecomment-1398334152)? If not, I figured it might be useful to make the node tfjs thing build.

Yes I did, but I would like something reproducible. It must be possible.

beardhatcode commented:
Finally got it working, now to clean it up and rebase on master

![image](https://github.com/user-attachments/assets/8271ab2d-96ac-4b42-b712-6b240ce44f52)

beardhatcode committed:
nextcloud: add ffmpeg to recognize for recognizing video and audio

beardhatcode commented:
The recognition of faces works, video and music also (although the rope skipping tutorial I uploaded is 0.017 
''battle rope training")

![image](https://github.com/user-attachments/assets/1871e0ef-0351-4ab5-8fcf-60a60677eeed)


I did not rebase on master as there are some things broken there...

beardhatcode committed:
nextcloud: remove useless ...

beardhatcode commented on pkgs/servers/nextcloud/packages/apps/recognize.nix:
```diff
@@ -0,0 +1,103 @@
+      url = "https://github.com/nextcloud/recognize/archive/refs/tags/v${version}.tar.gz";
+      hash = "sha256-op1Fdm40d8V6V+rfne0ECr60xRjBYrBPrCD3kaNeFfY=";
+    })
+    (fetchurl {
+      inherit version;
+      url = "https://storage.googleapis.com/tensorflow/libtensorflow/libtensorflow-cpu-linux-x86_64-2.9.1.tar.gz";
+      hash = "sha256-f1ENJUbj214QsdEZRjaJAD1YeEKJKtPJW8pRz4KCAXM=";
+    })
```
This is the only thing I still want to fix, to make it compatible with other platforms

beardhatcode commented:
> > This means we don't need something like https://github.com/NuschtOS/nixos-modules/blob/da5286bc062adee0e0aaf2bd3b784b477c623422/modules/nextcloud.nix#L172-L183 but still https://github.com/NuschtOS/nixos-modules/blob/da5286bc062adee0e0aaf2bd3b784b477c623422/modules/nextcloud.nix#L210-L217 ?
> 
> To be clear: I'm very, very 👎 on doing _any_ build steps at runtime.
> 
> I expect the JS code to be built correctly and to have the settings patched by this package to point to other Nix packages (I.e. node, ffmpeg etc.). I looked through those super-questionable[0] "migration" "steps"[1] and I really don't want to rely on any of that.
The installDeps file is now no longer doing anything. Node is substituted in place and ffmpeg is symlinked

> 
> [0] to say the least [1] https://github.com/nextcloud/recognize/blob/main/lib/Migration/InstallDeps.php

britter commented:
@beardhatcode I've been following this PR since it's inception. Thank you so much for investing so much time into packaging this app for NixOS :heart:

Ma27 commented on pkgs/servers/nextcloud/packages/apps/recognize.nix:
```diff
@@ -0,0 +1,103 @@
+
+    (fetchurl {
+      inherit version;
+      url = "https://github.com/nextcloud/recognize/archive/refs/tags/v${version}.tar.gz";
```
Why do we have this download and the one above?

Ma27 commented on pkgs/servers/nextcloud/packages/apps/recognize.nix:
```diff
@@ -0,0 +1,103 @@
+    # Merge the app and the models from github
+    tar -xzpf "${builtins.elemAt srcs 0}" recognize;
+    tar -xzpf "${builtins.elemAt srcs 1}" recognize-${version}/models;
+    mv recognize-${version}/models recognize
```
I think we can also unpack the models right into that directory, right?

beardhatcode replied:
yes, done

Ma27 commented on pkgs/servers/nextcloud/packages/apps/recognize.nix:
```diff
@@ -0,0 +1,103 @@
+    mv recognize-${version}/models recognize
+
+    # Place the tensorflow lib at the right place for building
+    cd recognize/node_modules/@tensorflow/tfjs-node/deps
```
Apologies if I forgot about something we discussed before, but I'm curious, did you try to package the node_modules stuff here via Nix fetchers or is that not feasible?

(non-blocking, would like to understand the reasoning for now)

beardhatcode replied:
The problem is that the install scripts of the tfjs node_modules fetches random files from the internet (in this case the src[3]) and it needs to use node-gyp. I did not find any other package that manages to fix this properly. Only photoprism has something that might be interesting that we could look into (but I would like to deffer that to a future PR)

Ma27 commented on pkgs/servers/nextcloud/packages/apps/recognize.nix:
```diff
@@ -0,0 +1,103 @@
+    sed -i "/'node_binary'/s:'""':'Nix Controled':" recognize/lib/Service/SettingsService.php
+
+    substituteInPlace recognize/lib/**/*.php \
+      --replace-quiet "\$this->settingsService->getSetting('node_binary')" "'${nodejs}/bin/node'"
```
`--replace-fail` is more appropriate here so we notice on updates that we need to update the patching, no?

beardhatcode replied:
I just looked at the source of substitute in place and since most of the files do not read the settings, I think it would fail. (While reading I discovered that I could combine the multiple invocations into one)

I added a check to validate that we indeed removed all occurrences

Ma27 replied:
Oh right, there's a glob. OK yes fine.

> I added a check to validate that we indeed removed all occurrences

Di you push it or am I just not seeing it?

beardhatcode replied:
I just pushed it now https://github.com/NixOS/nixpkgs/pull/333545/commits/69b961b12733745f6bbebc02c8cbae293b479e6b

Ma27 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -1,11 +1,17 @@
- * search of your editor. ;)
- * Hint: ### starts category names.
- */
-{ lib, noSysDirs, config, overlays }:
```
accidental nixfmt?

beardhatcode replied:
Likely :sweat_smile:

beardhatcode replied:
Rebased on latest nixos-unstable and fixed that commit

Ma27 replied:
GitHub still shows a merge conflict though?

beardhatcode replied:
I rebased on master now, and the conflict is resolved

beardhatcode committed:
nextcloud30Packages.apps.recognize: ensure node_binary is replaced

beardhatcode committed:
nextcloud: refactor path out of all-packages

beardhatcode committed:
nextcloud30Packages.apps.recognize: simplify tar extractions

beardhatcode committed:
nextcloud30Packages.apps.recognize: explain tensorflow version

beardhatcode committed:
nextcloud30Packages.apps.recognize: make tensorflow optional

For non-supported systems or those who want to use WebAssembly

    pkgs.nextcloud30Packages.apps.recognize.override ({ useLibTensorflow = false; });

Ma27 commented on pkgs/servers/nextcloud/packages/apps/recognize.nix:
```diff
@@ -0,0 +1,109 @@
+  ffmpeg,
+
+  # Current derivation only suports linux-x86_64 (contributions welcome, without libTensorflow builtin webassembly can be used)
+  useLibTensorflow ? stdenv.isx86_64 && stdenv.isLinux,
```
What is the alternative to this? Will it be downloaded at runtime? Or just fail?
Perhaps it's also fine to just support x86_64-linux for now with meta.platforms.

beardhatcode replied:
If you set it to false only the wasm version will work, (as the patch phase disables the installation). TensorFlow only supports  64 bit machines anyway (it does also support darwin, but I have no such device to test with). 

It should work on all platforms, just slower in WASM mode :sweat_smile:

beardhatcode commented:
@Ma27 and @SuperSandro2000 , for me, this PR is complete (unless there is something I missed).

Ma27 commented:
Skimmed over it last week and it looks good so far.
Just need to find some time to try it out :)

beardhatcode commented:
Has been working for me on my server for a couple of weeks now

beardhatcode committed:
nextcloud30Packages.apps.recognize: 8.1.1->8.2.0

beardhatcode committed:
nextcloudXXPackages.apps.recognize: suport nextcloud 29 and 28

beardhatcode commented:
I added support for older Nextclouds as well now, see https://apps.nextcloud.com/apps/recognize#downloads

britter commented:
Can't wait to see this merged!

SuperSandro2000 commented on pkgs/servers/nextcloud/packages/apps/recognize.nix:
```diff
@@ -0,0 +1,136 @@
+    cd -
+
+    # Test tfjs returns exit code 0
+    ${nodejs}/bin/node src/test_libtensorflow.js
```
```suggestion
    node src/test_libtensorflow.js
```
Please move node to nativeBuildInputs

teutat3s replied:
`nodejs` still needs to be moved to `nativeBuildInputs`.

beardhatcode replied:
Done :slightly_smiling_face:

SuperSandro2000 commented on pkgs/servers/nextcloud/packages/apps/recognize.nix:
```diff
@@ -0,0 +1,136 @@
+    nodejs
+    nodejs.pkgs.node-pre-gyp
+    nodejs.pkgs.node-gyp
+    python311
```
IIRC sometimes we can use python3(12) by adding setuptools and/or distutils.

SuperSandro2000 commented on pkgs/servers/nextcloud/packages/apps/recognize.nix:
```diff
@@ -0,0 +1,136 @@
+      tar -xzpf "${builtins.elemAt srcs 2}" -C recognize/node_modules/@tensorflow/tfjs-node/deps
+    '';
+
+  patchPhase = ''
```
```suggestion
  postPatch = ''
```

SuperSandro2000 commented on pkgs/servers/nextcloud/packages/apps/recognize.nix:
```diff
@@ -0,0 +1,136 @@
+    # Replace all occurences of node (and check that we actually remved them all)
+    test "$(grep "get[a-zA-Z]*('node_binary'" recognize/lib/**/*.php | wc -l)" -gt 0
+    substituteInPlace recognize/lib/**/*.php \
+      --replace-quiet "\$this->settingsService->getSetting('node_binary')" "'${nodejs}/bin/node'" \
+      --replace-quiet "\$this->config->getAppValueString('node_binary', '""')" "'${nodejs}/bin/node'" \
+      --replace-quiet "\$this->config->getAppValueString('node_binary')" "'${nodejs}/bin/node'" \
+      --replace-quiet "\$this->config->getAppValue('node_binary', '""')" "'${nodejs}/bin/node'" \
+      --replace-quiet "\$this->config->getAppValue('node_binary')" "'${nodejs}/bin/node'"
```
```suggestion
      --replace-error "\$this->settingsService->getSetting('node_binary')" "'${nodejs}/bin/node'" \
      --replace-error "\$this->config->getAppValueString('node_binary', '""')" "'${nodejs}/bin/node'" \
      --replace-error "\$this->config->getAppValueString('node_binary')" "'${nodejs}/bin/node'" \
      --replace-error "\$this->config->getAppValue('node_binary', '""')" "'${nodejs}/bin/node'" \
      --replace-error "\$this->config->getAppValue('node_binary')" "'${nodejs}/bin/node'"
```

beardhatcode replied:
Indeed.

There is a test verifying the replacement, not all files need replacing, bu by doing it this way it should keep working for longer

beardhatcode committed:
nextcloud30Packages.apps.recognize: patchPhase -> postPatch

beardhatcode committed:
nextcloudXXPackages.apps.recognize: use node from path

beardhatcode committed:
nextcloudXXPackages.apps.recognize: use python3 rather than 311

beardhatcode committed:
nextcloud30Packages.apps.recognize: s/buildInputs/nativeBuildInputs/

SuperSandro2000 commented:
@Ma27 we want to merge this?

dotlambda commented on pkgs/servers/nextcloud/packages/apps/recognize.nix:
```diff
@@ -0,0 +1,137 @@
+    cd recognize
+
+    # Install tfjs dependency
+    export CPPFLAGS="-I${nodejs}/include/node -Ideps/include"
```
```suggestion
    export CPPFLAGS="-I${lib.getDev nodejs}/include/node -Ideps/include"

dotlambda reviewed (commented):
`meta.changelog` and `meta.description` are missing

teutat3s commented on pkgs/servers/nextcloud/packages/apps/recognize.nix:
```diff
@@ -0,0 +1,137 @@
+in
+stdenv.mkDerivation rec {
+
+  pname = "nextcloud-app-recognise";
```
Typo snuck in here:
```
pname = "nextcloud-app-recognize";
```

teutat3s commented:
Would really like to backport this to 24.11.

Ma27 reviewed (commented):
```
error:
       … while evaluating an expression to select 'drvPath' on it
         at «internal»:1:552:
       … while calling the 'derivationStrict' builtin
         at «internal»:1:208:
       (stack trace truncated; use '--show-trace' to show the full trace)

       error: attribute 'node-pre-gyp' missing
       at /home/ma27/Projects/nixpkgs/pkgs/servers/nextcloud/packages/apps/recognize.nix:100:17:
           99|     nodejs
          100|     nodejs.pkgs.node-pre-gyp
             |                 ^
          101|     nodejs.pkgs.node-gyp
```

I'm sorry that it takes a while until I get to things sometimes, but did nobody even bother to evaluate this against latest master before merging?

EDIT: OK to be fair, this is a deprecated alias which makes me wonder why CI didn't complain. My money is on missing `recurseIntoAttrs` somewhere. But still, this needs to change.

beardhatcode commented:
I ran it on the latest master @Ma27 . Or at least that was what I thought I did, but maybe not. I did the following:

I rebased this PR on master, then I created and ran a vm for each of nextcloud{28,29,29} and I uploaded photos and videos and checked that it did what it should do. I also placed it on my server but that one is nixos-unstable.

Should I have done something additionally, or could you tell me how you bumped into this issue?

SomeoneSerge commented on pkgs/servers/nextcloud/packages/default.nix:
```diff
@@ -6,11 +6,12 @@
   lib,
   pkgs,
   newScope,
-  apps,
```
Hi! At a glance, this change seems to make it impossible to override the json manifests out-of-tree (I was using this in my overlays). If that's correct, here's a hotfix: https://github.com/NixOS/nixpkgs/pull/389371
